SA-485/bound llm response size

dstewartons opened the pull request:
# 📌 Pull Request Template

> **Please complete all sections**

## ✨ Summary

Fix 500 errors caused by oversized session cookies when LLM reasoning fields exceed browser cookie limits (4093 bytes). Implements conditional truncation of LLM reasoning text that only truncates to 500 characters when adding the reasoning would cause the session cookie to exceed the limit (3837 bytes with 256-byte safety margin). This preserves longer reasoning when the session has available space, only truncating when necessary to prevent cookie overflow errors that prevented survey responses from being saved.

## 📜 Changes Introduced

<!-- List key changes made in this PR. Consider bullet points for readability. -->

- [x] Bug fix (fix:) - Conditional truncation of LLM reasoning fields to prevent session cookie overflow
- [x] Updates to tests and/or documentation - Added integration test for dynamic truncation
- [ ] Terraform changes (if applicable)

**Key Changes:**
- Added `truncate_llm_reasoning_if_needed()` function in `utils/session_utils.py` that checks session size and only truncates reasoning to 500 characters when necessary
- Uses `get_encoded_session_size()` to calculate actual session cookie size before adding reasoning
- Truncates to fixed 500-character length when adding reasoning would exceed the 3837-byte limit (4093 bytes - 256-byte safety margin)
- Modified `add_classify_interaction()` to use conditional truncation instead of always truncating
- Added constants: `MAX_SESSION_COOKIE_SIZE` (4093), `SESSION_SIZE_SAFETY_MARGIN` (256), `MAX_ALLOWED_SESSION_SIZE` (3837), `REASONING_TRUNCATION_LENGTH` (500)
- Added integration test `test_add_classify_interaction_truncates_reasoning()` to verify session size stays within limits
- Truncation logs include person_id, original/truncated lengths, and calculated session sizes

**How Conditional Truncation Works:**
1. Creates a test session with the reasoning added and calculates what the session size would be using `get_encoded_session_size()`
2. Checks if the test size would exceed 3837 bytes (4093 - 256 safety margin)
3. If it fits: returns original reasoning unchanged (preserves longer reasoning when possible)
4. If it exceeds: truncates reasoning to 500 characters (fixed length)
5. Only truncates when necessary, preserving longer reasoning when the session has available space

## ✅ Checklist

> **Please confirm you've completed these checks before requesting a review.**

- [x] Code is formatted using **Black**
- [x] Imports are sorted using **isort**
- [x] Code passes linting with **Ruff**, **Pylint**, and **Mypy**
- [x] Security checks pass using **Bandit**
- [x] Tests are written and pass using **pytest**
- [x] Terraform files (if applicable) follow best practices and have been validated (`terraform fmt` & `terraform validate`)
- [x] DocStrings follow Google-style and are added as per Pylint recommendations
- [ ] Documentation has been updated if needed

## 🔍 How to Test

<!-- Describe how reviewers can verify your changes. Include test commands if applicable. -->

**Automated Tests:**
```bash
# Run all tests
make all-tests

# Run specific session utils truncation test
poetry run pytest tests/test_session_utils.py::test_add_classify_interaction_truncates_reasoning -v
```

**Manual Testing:**
1. Use the test case from SA-485 that reproduces the error:
   - Job title: "TikTok Affiliate"
   - Job description: "I review products and produce content for sellers on TikTok."
   - Organisation activity: "Marketing"
2. Verify that:
   - The survey completes successfully without 500 errors
   - Session cookie size stays under 3837 bytes (no "cookie is too large" warnings)
   - Survey responses are saved correctly
   - No cookie size warnings appear in logs

**Verification in GCP Logs:**

After entering the TikTok Affiliate test data, check logs for:
- **Truncation log** (if truncation occurred): 
  ```
  INFO - person_id:STPXXXXX-01 Reasoning would cause session to exceed limit (XXXX > 3837 bytes), truncating to 500 chars
  WARNING - person_id:STPXXXXX-01 LLM reasoning truncated from XXXX to 497 characters to prevent cookie overflow (session would be XXXX bytes, limit: 3837 bytes)
  ```
  The reasoning will be truncated to 500 characters (497 chars + "..." = 500 total) when session size would exceed the limit.
- **No truncation** (if session had enough space): Reasoning preserved at full length with no log messages (truncation only occurs when needed)
- **No cookie warnings**: Should NOT see `UserWarning: The 'session' cookie is too large`
- **No 500 errors**: Survey should complete successfully

**Key Indicators of Conditional Truncation:**
- Truncation to fixed 500 characters when session size would exceed 3837 bytes
- Truncation only occurs when necessary (not for all reasoning)
- Longer reasoning preserved when session has available space
- Logs show calculated session sizes and truncation details

dstewartons committed:
add tests for LLM reasoning truncation

Add comprehensive tests for truncate_llm_reasoning function and
integration test verifying truncation in add_classify_interaction.
Tests include request context setup for proper session access.

dstewartons committed:
truncate LLM reasoning to prevent cookie overflow

Add truncation logic to limit reasoning field to 500 characters before
saving to session, preventing 500 errors from oversized cookies.

dstewartons committed:
remove task ids

dstewartons committed:
reduce MAX_REASONING_LENGTH to 90

dlons commented:
The cookie size includes other session data (lookup responses with potential codes/divisions, follow-up questions, survey responses, metadata), so the reasoning field alone doesn't directly map 1:1 to cookie bytes. Think it's about 50 characters of reasoning ≈ 24-30 bytes of cookie size (accounting for JSON encoding and Base64 overhead). If a larger character limit for reasoning is desired (e.g., closer to 500 chars), we might need to also truncate other large session fields such as lookup response details (potential_codes, potential_divisions with long descriptions) to keep the total cookie size under 4093 bytes.

dstewartons committed:
consolidate truncation tests

fix lint error

remove comments

gibbardsteve reviewed (commented):
A suggestion to limit the truncation to only those cases where it is necessary.

dstewartons committed:
implement dynamic truncation for LLM reasoning
- calculate actual session size
- truncate when adding reasoning would exceed the 3837-byte limit (4093 - 256 safety margin)
- use binary search to preserve maximum reasoning length that fits within session limit.

dstewartons committed:
update truncation test for dynamic truncation implementation

dstewartons committed:
change to info

gibbardsteve commented on utils/session_utils.py:
```diff
@@ -418,6 +526,86 @@ def add_sic_lookup_interaction(
+
+    # Only truncate if adding reasoning would exceed the limit (4093 - 256 = 3837 bytes)
+    if test_size <= MAX_ALLOWED_SESSION_SIZE:
+        logger.info(
```
We don't want to log this for every survey response. Can you remove.

gibbardsteve reviewed (commented):
A comment about the binary search and to clean some logging.

dstewartons committed:
remove info log

dstewartons committed:
simplify model reasoning truncation to fixed 500 chars when session size would overflow

keep the size check using get_encoded_session_size but remove the binary search for max fitting length

dlons commented on utils/session_utils.py:
```diff
@@ -418,6 +526,86 @@ def add_sic_lookup_interaction(
+        f"({test_size} > {MAX_ALLOWED_SESSION_SIZE} bytes), truncating..."
+    )
+
+    best_length = _find_max_reasoning_length(reasoning, context)
```
PR updated

gibbardsteve reviewed (approved):
Looks good. Thanks.
